lib/tools/system-commands.js

```diff
@@ -80,71 +60,55 @@ systemCallMethods.connect = async function connect () {
 
   let isStartupMonitorEnabled = true;
   try {
-    if (_.isEmpty(await getPids(COMPANION_PGREP_PATTERN(this.udid)))) {
-      const args = buildCompanionArgs(this.udid, this.companion);
-      const comp = new SubProcess(IDB_COMPANION_EXECUTABLE, args);
-      comp.on('output', (stdout, stderr) => {
-        if (isStartupMonitorEnabled && stderr) {
-          log.debug(`[${IDB_COMPANION_EXECUTABLE}] ${stderr}`);
-        }
-      });
-      try {
-        await comp.start(null, PROCESS_INIT_TIMEOUT);
-      } catch (e) {
-        throw new Error(`Cannot start ${IDB_COMPANION_EXECUTABLE}. Original error: ${e.message}`);
-      }
-      await B.delay(300);
-      if (!comp.isRunning) {
-        throw new Error(`${IDB_COMPANION_EXECUTABLE} has failed to start`);
-      }
-      log.info(`${IDB_COMPANION_EXECUTABLE} has been started with args ${JSON.stringify(args)}`);
-    } else {
-      log.info(`${IDB_COMPANION_EXECUTABLE} is already running for the device '${this.udid}'`);
-    }
-
-    const daemon = new SubProcess(IDB_EXECUTABLE, buildDaemonArgs(this.executable));
-    let daemonStderr = '';
-    daemon.on('output', (stdout, stderr) => {
-      if (isStartupMonitorEnabled && stderr) {
-        daemonStderr += `${stderr}\n`;
-      }
-    });
-    try {
-      await daemon.start(null, PROCESS_INIT_TIMEOUT);
-      await B.delay(300);
-    } catch (ign) {}
-    if (daemon.isRunning) {
-      log.info(`${IDB_EXECUTABLE} daemon is running on port ${this.port}`);
-    } else {
-      if (!daemonStderr.includes('address already in use')) {
-        throw new Error(`${IDB_EXECUTABLE} daemon has failed to start: ${daemonStderr}`);
-      }
-      log.info(`${this.port} is already in use. Assuming it is used by ${IDB_EXECUTABLE} daemon`);
-    }
-
-    const {stdout} = await tpExec(IDB_EXECUTABLE, ['list-targets', '--json']);
-    if (!fixOutputToArray(stdout).some((x) => _.toUpper(x.udid) === _.toUpper(this.udid))) {
-      throw new Error(`The device '${this.udid}' is unknown. Only these devices are known ` +
-        `to ${IDB_EXECUTABLE}: ${stdout}`);
-    }
-
     try {
       await waitForCondition(async () => {
         try {
+          const daemon = new SubProcess(IDB_EXECUTABLE, buildDaemonArgs(this.executable));
+          let daemonOutput = '';
+          daemon.on('output', (stdout, stderr) => {
+            if (isStartupMonitorEnabled && _.trim(stdout || stderr)) {
+              daemonOutput += `[daemon] ${stdout || stderr}\n`;
+            }
+          });
+          try {
+            await daemon.start(null, PROCESS_INIT_TIMEOUT);
+            await B.delay(300);
```

Comment:
Since passing in a timeout to `daemon.start()` makes it return after that amount of time, why not just add `300` to the timeout and have it wait that much longer?

Reply:
according to the current teen_process logic the `start` call with given arguments will return as soon as the process writes anything into stdout/stderr or the timeout happens. This means that if the process writes an exception into stderr then `start` will exit as soon as the first stack trace line is printed, but the parent process itself will be still in running state to print the rest of the lines. This will make the following isRunning validation invalid